### ISSUE TITLE
Half-precision support for optimizers other than SGD

### ISSUE THREAD

**Issue by fhieber** [apache/incubator-mxnet#5384]
I would like to have some parameters of a network to be half-precision to save memory.
I am running into inconsistent behavior of different optimizers when using half-precision (float16) variables: for a toy network of y = softmax(w*x) with x float32 and w float16 cast to float32, the Adam optimizer sometimes(!) produces nan weights/scores, whereas SGD seems to work fine.
(Run example below with x_dtype = 'float32' and w_dtype = 'float16').
Other optimizers such as adagrad/adadelta don't seem to support float16 at all, leading to a attr check fail.

## Environment info
Operating System:
MacOS

MXNet version:
8b6c8598f8f15a765dd9695c92ce251cd9eda747

## Steps to reproduce

```python
import mxnet as mx
import logging
logging.basicConfig(level=logging.INFO)

x_dtype='float32'
w_dtype='float16'
optimizer = 'adam'

x = mx.sym.Variable('data', dtype=x_dtype)
label = mx.sym.Variable('softmax_label', dtype=x_dtype)
w = mx.sym.cast(mx.sym.Variable("weight", dtype=w_dtype), dtype=x_dtype)
y = mx.sym.FullyConnected(x, w, num_hidden=10)
y = mx.sym.SoftmaxOutput(data=y, label=label)
mod = mx.module.Module(y)
it = mx.io.NDArrayIter(mx.nd.uniform(0,1,(10,10), dtype=x_dtype), mx.nd.ones((10,), dtype=x_dtype))
mod.fit(it, num_epoch=10, eval_metric='ce', optimizer=optimizer, optimizer_params={'learning_rate': 0.00001})
```

x_dtype | w_dtype | optimizer | outcome
------------ | ------------- | ------------- | -------------
float32 | float16 | adam | nan (sometimes)
float32 | float16 | adagrad/adadelta | check fail
float32 | float32 | * | ok
float16 | float16 | * | FC float16 only supported by CuDNN

What is the current support of optimizers for these scenarios? If not supported at all, is Adam simply missing a check like AdaGrad/AdaDelta?

**Comment by fhieber** [comment]
Also, is it planned to have full support of float16 in general? For example, FullyConnected with float16 data seems to only work on GPUs/cuDNN.

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
